source/fe/fe_system.cc

```diff
@@ -42,8 +42,185 @@ namespace
     });
   }
 } // namespace
-/* ----------------------- FESystem::InternalData ------------------- */
 
+namespace internal
+{
+  /**
+   * Setup a table of offsets for a primitive FE. Unlike the nonprimitive
+   * case, here the number of nonzero components per shape function is always
+   * 1 and the number of components in the FE is always the multiplicity.
+   */
+  template <int dim, int spacedim = dim>
+  Table<2, unsigned int>
+  setup_primitive_offset_table(const FESystem<dim, spacedim> &fe,
+                               const unsigned int             base_no)
+  {
+    Assert(fe.base_element(base_no).is_primitive(), ExcInternalError());
+    Table<2, unsigned int> table(fe.element_multiplicity(base_no),
+                                 fe.base_element(base_no).n_dofs_per_cell());
+    // 0 is a bad default value since it is a valid index
+    table.fill(numbers::invalid_unsigned_int);
+
+    for (unsigned int system_index = 0; system_index < fe.n_dofs_per_cell();
+         ++system_index)
+      if (fe.system_to_base_index(system_index).first.first == base_no)
+        {
+          Assert(fe.n_nonzero_components(system_index) == 1,
+                 ExcInternalError());
+          const unsigned int base_component =
+            fe.system_to_base_index(system_index).first.second;
+          const unsigned int base_index =
+            fe.system_to_base_index(system_index).second;
+          Assert(base_index < fe.base_element(base_no).n_dofs_per_cell(),
+                 ExcInternalError());
+
+          unsigned int out_index = 0;
+          for (unsigned int i = 0; i < system_index; ++i)
+            out_index += fe.n_nonzero_components(i);
```

Comment:
While you only pay this cost at setup, I still think we should avoid this second loop (making this a quadratic complexity algorithm). I think you could avoid this by defining `out_index` before line 64 and incrementing it by `fe.n_nonzero_components(system_index)` after the use in line 80.

Reply:
Fixed!